File: src_local/utils/indexer.py
"""
SISTEMA DE PERSONAS AGENTES - INDEXADOR TÉCNICO
Módulo: Indexador de Metadados (Indexer)
Função: Gerar o mapa de conhecimento estrutural (classes/funções).
Soberania: CORE-UTILITY.
"""
import logging
import ast
import time
from pathlib import Path

logger = logging.getLogger(__name__)

class Indexer:
    """
    Estrategista de Conhecimento: Cataloga a estrutura anatômica do projeto 📑.
    
    Responsabilidades:
    1. Varredura Estrutural: Identifica todos os componentes Python do ecossistema.
    2. Extração de Consciência: Mapeia classes e funções para o grafo de chamadas.
    3. Persistência de Conhecimento: Mantém o 'index.json' sincronizado.
    """
    
    def __init__(self, project_root):
        self.project_root = Path(project_root)
        self.index_file = self.project_root / "index.json"

    def update_index(self):
        """
        📑 Atualiza o índice global de metadados técnicos do projeto.
        Realiza uma leitura profunda via AST de todos os módulos Python para
        atualizar a base de conhecimento estrutural utilizada pelo Maestro.
        """
        logger.info("📡 Iniciando indexação soberana de metadados...")
        index_data = {"last_update": time.strftime('%Y-%m-%d %H:%M:%S'), "files": {}}
        
        try:
            for path in self.project_root.rglob("*.py"):
                # Veto de Escopo: Ignora metadados de controle e testes
                if any(part in ["__pycache__", "build", "tests"] for part in path.parts):
                    continue
                
                rel_path = path.relative_to(self.project_root).as_posix()
                index_data["files"][rel_path] = self._extract_metadata(path)
                
            logger.info(f"✅ Indexação concluída: {len(index_data['files'])} módulos catalogados.")
            return index_data
        except Exception as e:
            logger.error(f"🚨 Falha crítica no motor de indexação: {e}", exc_info=True)
            return index_data

    def _extract_metadata(self, path: Path):
        """
        🧬 Extrai a anatomia do módulo via análise AST.
        Mapeia a estrutura de classes e funções para a base de conhecimento.
        Garante a rastreabilidade da topologia sistêmica.
        """
        try:
            import time
            start_ast = time.time()
            content = path.read_text(encoding='utf-8', errors='ignore')
            tree = ast.parse(content)
            res = {
                "classes": [node.name for node in ast.walk(tree) if isinstance(node, ast.ClassDef)],
                "functions": [node.name for node in ast.walk(tree) if isinstance(node, ast.FunctionDef)]
            }
            duration = time.time() - start_ast
            if duration > 0.05:
                logger.debug(f"⏱️ [Indexer] AST extraída em {path.name}: {duration:.4f}s")
            return res
        except Exception as e:
            # Veto de Silenciamento: Reporta falha na decomposição para o log forense
            logger.warning(f"⚠️ Falha na anatomia AST de {path.name}: {e}")
            return {"error": str(e)}

File: src_local/utils/indexer.ts
import winston from "winston";
import { HubManagerGRPC } from "../core/hub_manager_grpc.ts";

const logger = winston.child({ module: "Indexer" });

export interface FileMetadata {
    classes: string[]; functions: string[]; exports: string[]; lines: number; error?: string;
}

export interface IndexData {
    lastUpdate: string; files: Record<string, FileMetadata>;
    stats: { totalFiles: number; totalClasses: number; totalFunctions: number; totalExports: number; };
}

/**
 * 📑 Indexer — Cataloga a estrutura anatômica do projeto via gRPC Proxy (Go Hub).
 */
export class Indexer {
    constructor(private projectRoot: string, private hubManager?: HubManagerGRPC) { }

    async updateIndex(): Promise<IndexData> {
        logger.info("📡 Iniciando indexação soberana via Hub Proxy (gRPC)...");
        const startTime = Date.now();
        const indexData = this.getEmptyIndex();

        if (!this.hubManager) {
            logger.warn("⚠️ HubManager not provided to Indexer. Returning empty index.");
            return indexData;
        }

        try {
            const processedFiles = await this.hubManager.indexProject(this.projectRoot);

            if (!processedFiles || !Array.isArray(processedFiles)) {
                throw new Error(`Invalid response from Hub indexer.`);
            }

            for (const pf of processedFiles) {
                const metadata: FileMetadata = {
                    classes: pf.classes || [],
                    functions: pf.functions || [],
                    exports: pf.exports || [],
                    lines: pf.lines || 0
                };
                this.integrateResult(indexData, pf.path, metadata);
            }

            const duration = ((Date.now() - startTime) / 1000).toFixed(2);
            logger.info(`✅ Indexação nativa concluída: ${indexData.stats.totalFiles} módulos em ${duration}s.`);
        } catch (e: any) {
            logger.error(`🚨 Falha crítica no motor de indexação: ${e.message}`);
        }
        return indexData;
    }

    private getEmptyIndex(): IndexData {
        return {
            lastUpdate: new Date().toISOString(), files: {},
            stats: { totalFiles: 0, totalClasses: 0, totalFunctions: 0, totalExports: 0 },
        };
    }

    private integrateResult(indexData: IndexData, file: string, metadata: FileMetadata) {
        indexData.files[file] = metadata;
        indexData.stats.totalFiles++;
        indexData.stats.totalClasses += metadata.classes.length;
        indexData.stats.totalFunctions += metadata.functions.length;
        indexData.stats.totalExports += metadata.exports.length;
    }
}

